### PR TITLE
Create auxilary files for game edition/expansion information

### LINKED ISSUE #1314: Store game edition and game expansion info in auxiliary file instead of hardcoding it
**Required skills**: Python

**Difficulty**: Easy

openage currently stores information about a game edition or expansion inside an `Enum`, e.g. media paths, support status, etc. We should store this information in an auxiliary file instead, similar to #1312. That would make updating and correcting game information easier as we no longer have to replace chunks of code.

The recommended approach is to transfer the information from the existing implementation into a simple data definition file format. TOML is prefered because we already use that elsewhere in the converter.

Example:

```toml
# Header
# The version of this config file format
file_version = "1.0"

# Game edition info
game_edition_id = "AOC"
support_status = True

[media_paths]
datfile = [ "data/empires2.dat", ]
graphics = [ "data/graphics.drs", ]
palettes = [ "data/interfac.drs", ]
```

Since we cannot use Enums when we read the game versions from file, we would have to replace identity checks of the game version `Enum` value with an ID check. These checks happen when reading the `.dat` format or in the service that acquires nyan object names. They should be easy to replace.

**Further Reading**

* [GameVersion(Enum) implementation](https://github.com/SFTtech/openage/blob/master/openage/convert/value_object/init/game_version.py)

### PR DESCRIPTION
**Pull request by drs-11**
Fixes #1314 and #1312.

Things done so far:
- created version-hash toml files for game editions and expansions inside `version-hashes` directory
   - every game/expansion has a single toml file
   - remaining games with MD5 hashes were also created(not sure if this was needed to be done)
   - a directory was created to store the files
- created auxiliary files from `GameEdition` and `GameExpansion` enums
   - single files were created for both

@heinezen any changes or additions needs to be done?

Also `AOE1DE` enum in `game_version.py` has same MD5 hashes for both `AoEDE_s.exe` and `empires.dat`. Is that a coincidence?

### CODE REVIEW

**Comment by heinezen** [comment]
The TOML files look good :rocket: 

Now there are a few additional changes we have to make:

* I have updated the list of SHA3 hashes in https://github.com/SFTtech/openage/issues/1312#issuecomment-715425363 to include all game versions now. You should be able to exclusively use SHA3 with this.
* All TOML files need to move outside of the converter code into a "user area" where we can easily update them. A good place in the repo would be `cfg/converter/games/`
* Since we will read the game versions from a file instead of hardcoding them, we obviously can't let them be `Enum`s anymore.
    * `GameEdition` and `GameExpansion` need to be turned into regular classes (should be trivial)
    * These objects now have to be created by reading the files with `toml` and initializing them with the data inside. This can happen in a function inside https://github.com/SFTtech/openage/blob/master/openage/convert/service/init/version_detect.py . The function should return 
        1. A list of `GameEdition` objects
        2. A list of `GameExpansions` objects
    * They should get a new member `game_id` which is initialized with `game_edition_id` from your TOML files
    * These objects should be stored in `args.avail_game_eds` and `args.avail_game_exps`, since we need them later in the conversion process
* All functions that previously used the enums have to be adapted. The most important places are:
    1. The [version detection](https://github.com/SFTtech/openage/blob/e6879138514ea982870681517167bfc5d76b1149/openage/convert/service/init/version_detect.py#L11) service. This needs to get passed the games versions stored in `args`
    2. Every file in the [dat format reader](https://github.com/SFTtech/openage/tree/master/openage/convert/value_object/read/media/datfile). Instead of checking for identity, all version checks should be turned in string cmparisons based on `game_id`. As a result

    ```
    if game_version[0] is GameEdition.ROR:
    ```

    should become

    ```
    if game_version[0].game_id == "ROR":
    ```
    3. Other files might reference `GameEdition` or `GameExpansion` and need to be updated like in ii. You have to search the repo to find them

I can provide you a version of AoC for testing if you need it, btw. Write me in the chat for details :)

> Also AOE1DE enum in game_version.py has same MD5 hashes for both AoEDE_s.exe and empires.dat. Is that a coincidence?

Probably a copy-paste error on my behalf :D

**Comment by drs-11** [comment]
@heinezen need a little help in saving the objects to `args`.

I thought calling the `create_version_objects` function in `convert_assets` function would be a good idea but having a little trouble accessing the cfg folder where the auxiliary files are located. I saw somewhere `args.cfg_dir` being used, can I use that?

**Comment by heinezen** [comment]
@drs-11 You can assign `root["cfg"]` directly to `args.cfg_dir`, since `args` is a namespace. For example

```
args.cfg_dir = root["cfg"]
```
somewhere in `convert.main.py`. The occurence of `args.cfg_dir` you saw is probably in another subparser.

File: openage/convert/value_object/init/game_version.py [hunk 1]
```diff
@@ -22,75 +22,92 @@ class Support(enum.Enum):
...
+                      for the converter.
+        """
+        self.game_id = game_id
+        self.flags = flags
+
+        self.support_status = None
+        self.game_file_versions = []
+        self.media_paths = {}
+        self.target_modpacks = None
+        self.expansions = []
```
Comment (heinezen) [inline openage/convert/value_object/init/game_version.py:44]: Game edition/expansion objects should be immutable (i.e. they should not be altered during conversion for obvious reasons) and therefore, all members should be set during object initialization. You can still use the functions you implemented, but they should be private and only be called from `__init__()`. This means you also have to pass it all necessary values as parameters.

In our architecture view, game editions/expansions are [value objects](https://github.com/SFTtech/openage/blob/master/doc/code/converter/architecture_overview.md#value-object).

**Comment by drs-11** [comment]
@heinezen enum changes are almost finished. Have few queries on remaining enums.

File `struct_definition.py` calls `GameEdition.AOC` directly for `target.get_data_format` function. Which doesn't really uses the enum from what I've looked up. Is that right? Then is it safe to change it to string "AOC" ?

File `singlefile.py` uses `GameEdition.AOC` enum for creating DRS objects. I think passing string "AOC" will require a few changes in other places which calls DRS as well. Should I go ahead with this?

File: openage/convert/main.py [hunk 2]
```diff
@@ -175,8 +180,14 @@ def main(args, error):
     from ..assets import get_asset_path
     outdir = get_asset_path(args.output_dir)
 
+    # mount the config folder at "cfg/"
```
Comment (drs-11) [inline openage/convert/main.py:183]: Umm not sure I did it right here. 

There's also `interactive_browser` function at line 175 which requires the `args` to have the `GameEdition` and `GameExpansion` objects to be available. I should push it after `convert_assets` function right?

**Comment by heinezen** [comment]
> File struct_definition.py calls GameEdition.AOC directly for target.get_data_format function. Which doesn't really uses the enum from what I've looked up. Is that right? Then is it safe to change it to string "AOC" ?

Yes, you just give it the `game_id` identifier. `struct_definition.py` is a legacy module btw, so you don't need to worry about it much after changing the parameter to string. If the build fails because of that I will fix it myself.

> File singlefile.py uses GameEdition.AOC enum for creating DRS objects. I think passing string "AOC" will require a few changes in other places which calls DRS as well. Should I go ahead with this?

I think you can do that. The init function getting a string is fine for now. I think the only other place that reads the DRS is the main converter in this place: https://github.com/SFTtech/openage/blob/64b8c740ac3cf97131470a15bded56bf6ab97095/openage/convert/service/init/mount_asset_dirs.py#L27 . You should be able to easily change that.

File: openage/convert/main.py (hunk 2 above)
Comment (heinezen) [reply openage/convert/main.py:183]: No, the interactive browser is a separate module that can be invoked before conversion. I would recommend you also pass it the cfg dir and replicate the game edition/expansion reading and subsequent version detection to the interactive browser.

File: openage/convert/main.py [hunk 3]
```diff
@@ -35,8 +36,12 @@ def convert_assets(assets, args, srcdir=None, prev_source_dir_path=None):
     if srcdir is None:
         srcdir = acquire_conversion_source_dir(prev_source_dir_path)
 
+    # Initialize game versions data
+    auxiliary_files_dir = cfg / "converter" / "games"
+    args.avail_game_eds, args.avail_game_exps = create_version_objects(auxiliary_files_dir)
+
     # Acquire game version info
-    game_version = get_game_version(srcdir)
+    game_version = get_game_version(srcdir, args)
```
Comment (heinezen) [inline openage/convert/main.py:44]: You do not need to pass all of `args` here. Only the list of game editions and the list of game expansions needs to be passed.

File: openage/convert/value_object/read/media/datfile/sound.py [hunk 4]
```diff
@@ -21,13 +20,13 @@ def get_data_format_members(cls, game_version):
...
 
-        if game_version[0] in (GameEdition.AOE1DE, GameEdition.AOE2DE):
+        if game_version[0].game_id in ("AOE1DE", "AOE2DE"):
             data_format.extend([
                 (SKIP, "name_len_debug", StorageType.INT_MEMBER, "uint16_t"),
                 (READ, "name_len", StorageType.INT_MEMBER, "uint16_t"),
                 (READ_GEN, "name", StorageType.STRING_MEMBER, "char[name_len]"),
             ])
-        elif game_version[0] is GameEdition.SWGB:
+        elif game_version[0].game_id is "SWGB":
```
Comment (heinezen) [inline openage/convert/value_object/read/media/datfile/sound.py:29]: Should be `==`

**Comment by drs-11** [comment]
Getting `ModuleNotFoundError` for `toml` in `version_detect.py`. Hmmm weird.
Not sure why it's happening.

**Comment by heinezen** [comment]
@TheJJ Can you check if kevin installs toml correctly?

**Comment by TheJJ** [comment]
Done, toml is now in the image. But unfortunately this also means pylint 1.6, and a lot of new compliance issues.....

**Comment by drs-11** [comment]
Ok yeah there's tons of linting issues now which weren't there previously. Also putting string instead of object in `struct_definition.py` led to some implemented `get_data_format_members` functions to have errors. I guess just need to modify the `game_id` checking in those functions.

**Comment by drs-11** [comment]
@heinezen I think I forgot to mention that I passed string "AOC" in `GenieStructure` class as well. The inherited classes in their `get_data_format_members` function will need to have their `game_id` checking as a string instead of objects. Should I go ahead?

**Comment by heinezen** [comment]
@drs-11 I will deal with the legacy stuff myself. Its probably to frustrating for you to deal with.

**Comment by drs-11** [comment]
@heinezen alright then. Anything else that needs to be done in this issue?

**Comment by heinezen** [comment]
@drs-11 Have you ticket the box "Allow edits from maintainers" on the right?

**Comment by drs-11** [comment]
@heinezen yup it's ticked.

**Comment by heinezen** [comment]
@drs-11 For some reason, I can't push, so please extract and apply this patch to fix the buildsystem:


[0001-convert-Fix-buildsystem-requirement-for-new-GameVers.zip](https://github.com/SFTtech/openage/files/5475948/0001-convert-Fix-buildsystem-requirement-for-new-GameVers.zip)

**Comment by heinezen** [comment]
It builds again! Now its time for the review

File: openage/convert/service/conversion/internal_name_lookups.py [hunk 5]
```diff
@@ -90,13 +89,13 @@ def get_class_lookups(game_version):
     game_edition = game_version[0]
     # game_expansions = game_version[1]
 
-    if game_edition is GameEdition.ROR:
+    if game_edition.game_id == "ROR":
         return ror_internal.CLASS_ID_LOOKUPS
 
-    if game_edition in (GameEdition.AOC, GameEdition.HDEDITION, GameEdition.AOE2DE):
+    if game_edition in ("AOC", "HDEDITION", "AOE2DE"):
```
Comment (heinezen) [inline openage/convert/service/conversion/internal_name_lookups.py:95]: Change to: `game_edition.game_id`

File: openage/convert/service/conversion/internal_name_lookups.py [hunk 6]
```diff
@@ -204,13 +203,13 @@ def get_gather_lookups(game_version):
     game_edition = game_version[0]
     # game_expansions = game_version[1]
 
-    if game_edition is GameEdition.ROR:
+    if game_edition.game_id == "ROR":
         return ror_internal.GATHER_TASK_LOOKUPS
 
-    if game_edition in (GameEdition.AOC, GameEdition.HDEDITION, GameEdition.AOE2DE):
+    if game_edition in ("AOC", "HDEDITION", "AOE2DE"):
```
Comment (heinezen) [inline openage/convert/service/conversion/internal_name_lookups.py:209]: Change to: `game_edition.game_id`

File: openage/convert/service/conversion/internal_name_lookups.py [hunk 7]
```diff
@@ -260,13 +259,13 @@ def get_restock_lookups(game_version):
     game_edition = game_version[0]
     # game_expansions = game_version[1]
 
-    if game_edition is GameEdition.ROR:
+    if game_edition.game_id == "ROR":
         return None
 
-    if game_edition in (GameEdition.AOC, GameEdition.HDEDITION, GameEdition.AOE2DE):
+    if game_edition in ("AOC", "HDEDITION", "AOE2DE"):
```
Comment (heinezen) [inline openage/convert/service/conversion/internal_name_lookups.py:265]: Change to: `game_edition.game_id`

**Review by heinezen — CHANGES_REQUESTED** [review]
For next time: **Please test your code.** A lot of these bugs can be avoided by doing a test run.

**Comment by drs-11** [comment]
Omg I'm really sorry, I had been thinking CI tests would be enough for this issue. Clearly mistaken. Sorry for all the trouble!

**Comment by drs-11** [comment]
Tested with AoC this time. Sorry for the inconvenience caused.

**Comment by heinezen** [comment]
@drs-11 Don't worry, it was not that troubling :)

The reason I was emphasizing testing is that it becomes very hard to write a good review if the code does not run because

1. I have to get it running by second-guessing the correct code that you wanted to write
2. Debug my own fixes and your code simultaneously

This can be very confusing :D

**Review by heinezen — DISMISSED** [review]
Runs through for

* AoC
* RoR
* SWGB

So it should be fine.

**Comment by heinezen** [comment]
@drs-11 Plz do a rebase, then we can merge :)

**Comment by drs-11** [comment]
> The reason I was emphasizing testing is that it becomes very hard to write a good review

Absolutely, I agree.

> Plz do a rebase

Squash the commits as well then?

**Comment by heinezen** [comment]
@drs-11 No squash, only rebase

**Comment by drs-11** [comment]
@heinezen done

**Comment by drs-11** [comment]
@heinezen done

File: openage/convert/service/init/version_detect.py [hunk 8]
```diff
@@ -47,3 +54,78 @@ def iterate_game_versions(srcdir):
...
+    version_hashes_path = srcdir.joinpath("version-hashes")
+
+    # load toml config files to a dictionary variable
+    with game_edition_path.open() as game_edition_toml:
+        game_edition_dic = toml.loads(game_edition_toml.read())
+    with game_expansion_path.open() as game_expansion_toml:
+        game_expansion_dic = toml.loads(game_expansion_toml.read())
+
+    # create and list GameEdition objects
+    file_version_game_ed = game_edition_dic.pop("file_version")
```
Comment (heinezen) [inline openage/convert/service/init/version_detect.py:80]: We can fix the sanity_check test here by changing the line to just

```
game_edition_dic.pop("file_version")
```

File: openage/convert/service/init/version_detect.py [hunk 9]
```diff
@@ -47,3 +54,78 @@ def iterate_game_versions(srcdir):
...
+    for game in game_edition_dic:
+        game_obj = create_game_obj(game_edition_dic[game], version_hashes_path)
+        game_edition_list.append(game_obj)
+
+    # create and list GameExpansion objects
+    file_version_game_ex = game_expansion_dic.pop("file_version")
```
Comment (heinezen) [inline openage/convert/service/init/version_detect.py:86]: We can fix the sanity_check test here by changing the line to just

```
game_expansion_dic.pop("file_version")
```

**Comment by heinezen** [comment]
@TheJJ Any idea why the sanity check reports the same issue multiple times?

**Review by heinezen — APPROVED** [review]

**Comment by heinezen** [comment]
Nice :)

**Comment by drs-11** [comment]
_Phew_ finally.
Thanks! Definitely learned a few things doing this issue :D

**Comment by heinezen** [comment]
Definitely, that was two tasks/issues at once :D

**Pull request merged**